core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrRowPatternFlattener.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.rowpattern;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+/**
+ * Optimize row pattern:
+ * - unfold grouping
+ * - remove nested exclusions
+ * - flatten alternations and concatenations
+ * - remove redundant empty pattern
+ */
+public final class IrRowPatternFlattener
+        extends IrRowPatternVisitor<IrRowPattern, Boolean>
+{
+    public static IrRowPattern optimize(IrRowPattern node)
+    {
+        return new IrRowPatternFlattener().process(node, false);
+    }
+
+    @Override
+    protected IrRowPattern visitIrRowPattern(IrRowPattern node, Boolean inExclusion)
+    {
+        throw new UnsupportedOperationException("unsupported node type: " + node.getClass().getName());
+    }
+
+    @Override
+    protected IrRowPattern visitIrLabel(IrLabel node, Boolean inExclusion)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrAnchorPattern(IrAnchorPattern node, Boolean inExclusion)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrEmptyPattern(IrEmptyPattern node, Boolean inExclusion)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrExcludedPattern(IrExcludedPattern node, Boolean inExclusion)
+    {
+        IrRowPattern child = process(node.getPattern(), true);
+        if (inExclusion) {
+            // skip nested exclusion. this is necessary to resolve exclusions correctly during pattern matching
+            return child;
+        }
+
+        return new IrExcludedPattern(child);
+    }
+
+    @Override
+    protected IrRowPattern visitIrPatternAlternation(IrPatternAlternation node, Boolean inExclusion)
```

Comment:
Can you explain what the logic in this method does? Not sure I follow the intent. It would be great if there were a few examples of what gets transformed into what.

Reply:
I added a comment with some examples.
There are more examples in `TestIrRowPatternOptimization`, but they are not so easy to follow after I changed the tests to start from IrRowPatter instead of parse String pattern.